### PR TITLE
Tau HLT L1 match update

### PR DESCRIPTION
**Pull request by tstreble**
This PR:
- introduces a new CaloTower producer for the regional PFlow based on L1 Tau objects associated with a fired L1 seed

### CODE REVIEW

File: RecoTauTag/HLTProducers/src/CaloTowerFromL1TSeededCreatorForTauHLT.cc [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+  mtowers_token     (consumes<CaloTowerCollection>    (p.getParameter<InputTag>     ("towers"            ))),
+  mCone                                               (p.getParameter<double>       ("UseTowersInCone"   ) ),
+  mTauTrigger_token (consumes<trigger::TriggerFilterObjectWithRefs>   (p.getParameter<InputTag>     ("TauTrigger"        ))),
+  mEtThreshold                                        (p.getParameter<double>       ("minimumEt"         ) ),
+  mEThreshold                                         (p.getParameter<double>       ("minimumE"          ) )
+{
+  produces<CaloTowerCollection>();
+}
+
+CaloTowerFromL1TSeededCreatorForTauHLT::~CaloTowerFromL1TSeededCreatorForTauHLT() {
```
Comment (fwyzard) [inline RecoTauTag/HLTProducers/src/CaloTowerFromL1TSeededCreatorForTauHLT.cc]: please replace this with
`CaloTowerFromL1TSeededCreatorForTauHLT::~CaloTowerFromL1TSeededCreatorForTauHLT() = default;`

File: RecoTauTag/HLTProducers/interface/CaloTowerFromL1TSeededCreatorForTauHLT.h [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  ~CaloTowerFromL1TSeededCreatorForTauHLT();
+  /// 
+  static void fillDescriptions( edm::ConfigurationDescriptions& desc );
+
+ private:
+  /// process one event
+  void produce( edm::StreamID sid, edm::Event& evt, const edm::EventSetup& stp ) const override;
+
+  /// verbosity
+  const int mVerbose;
```
Comment (fwyzard) [inline RecoTauTag/HLTProducers/interface/CaloTowerFromL1TSeededCreatorForTauHLT.h]: please change the data members to follow the CMSSW naming conventions: either `m_verbose` or `verbose_`, and similarly for the others

File: RecoTauTag/HLTProducers/src/CaloTowerFromL1TSeededCreatorForTauHLT.cc [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+void CaloTowerFromL1TSeededCreatorForTauHLT::produce( StreamID sid, Event& evt, const EventSetup& stp ) const {
+  edm::Handle<CaloTowerCollection> caloTowers;
+  evt.getByToken( mtowers_token, caloTowers );
+
+  // L1 seeds
+  edm::Handle<trigger::TriggerFilterObjectWithRefs> l1TriggeredTaus;
+  evt.getByToken( mTauTrigger_token, l1TriggeredTaus );
+
+  std::unique_ptr<CaloTowerCollection> cands( new CaloTowerCollection );
```
Comment (fwyzard) [inline RecoTauTag/HLTProducers/src/CaloTowerFromL1TSeededCreatorForTauHLT.cc]: please use
```c++
auto cands = std::make_unique<CaloTowerCollection>();
```

File: RecoTauTag/HLTProducers/src/CaloTowerFromL1TSeededCreatorForTauHLT.cc [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+  cands->reserve( caloTowers->size() );
+
+  l1t::TauVectorRef tauCandRefVec;
+  l1TriggeredTaus->getObjects( trigger::TriggerL1Tau,tauCandRefVec);
+
+  for(unsigned int iL1Tau = 0; iL1Tau < tauCandRefVec.size(); iL1Tau++){
```
Comment (fwyzard) [inline RecoTauTag/HLTProducers/src/CaloTowerFromL1TSeededCreatorForTauHLT.cc]: please use
```c++
for(auto const& tauCandRef: tauCandRefVec){
```
and change `tauCandRefVec[iL1Tau]` to `tauCandRef` below.

**Review by fwyzard — CHANGES_REQUESTED** [review]

**Comment by fwyzard** [comment]
@cmsbuild, please test

**Comment by Martin-Grunewald** [comment]
please test

File: RecoTauTag/HLTProducers/src/CaloTowerFromL1TSeededCreatorForTauHLT.cc [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+  auto cands = std::make_unique<CaloTowerCollection>();
+  cands->reserve( caloTowers->size() );
+
+  l1t::TauVectorRef tauCandRefVec;
+  l1TriggeredTaus->getObjects( trigger::TriggerL1Tau,tauCandRefVec);
+
+  for(auto const& tauCandRef: tauCandRefVec){
+
+    unsigned idx   = 0 ;
+    for (; idx < caloTowers->size(); idx++) {
```
Comment (davidlange6) [inline RecoTauTag/HLTProducers/src/CaloTowerFromL1TSeededCreatorForTauHLT.cc]: hi @tstreble - please use a range based loop here

File: RecoTauTag/HLTProducers/src/CaloTowerFromL1TSeededCreatorForTauHLT.cc [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+	edm::LogInfo("JetDebugInfo") << "CaloTowerFromL1TSeededCreatorForTauHLT::produce-> " << idx 
+				     << " tower et/eta/phi/e: "                        << cal->et()  << '/' 
+				     << cal->eta() << '/' 
+				     << cal->phi() << '/' 
+				     << cal->energy() 
+				     << " is...";
+      }
+      if (cal->et() >= m_EtThreshold && cal->energy() >= m_EThreshold ) {
+	math::PtEtaPhiELorentzVector p( cal->et(), cal->eta(), cal->phi(), cal->energy() );
+	double delta  = ROOT::Math::VectorUtil::DeltaR((tauCandRef->p4()).Vect(), p);
```
Comment (davidlange6) [inline RecoTauTag/HLTProducers/src/CaloTowerFromL1TSeededCreatorForTauHLT.cc]: please use DataFormats/Math/interface/deltaR (or better deltaR2)

**Comment by Martin-Grunewald** [comment]
please test

**Comment by Martin-Grunewald** [comment]
+1

**Comment by davidlange6** [comment]
merge

**Pull request merged**